src/protocol.rs

```diff
@@ -425,35 +432,42 @@ fn setup_query_context(
 }
 
 /// Encrypts a message `msg` using the symmetric key derived from query
+/// An optional `response_nonce` can be specified that will be used to derive the encryption keys
+/// If `response_nonce` is not specified, it will be randomly generated
 /// In practice, this is used to encrypt the response body before sending
 /// to client
 async fn encrypt_response_helper(
     odoh_secret: &[u8],
     plaintext_resp_body: &[u8],
+    response_nonce: Option<Vec<u8>>,
     query: &ObliviousDoHQueryBody,
-) -> Result<Vec<u8>> {
-    let (key, nonce) = derive_secrets(odoh_secret, query)?;
+) -> Result<(Vec<u8>, ResponseNonce)> {
+    let (key, nonce, response_nonce) = derive_secrets(odoh_secret, query, response_nonce)?;
     let cipher = Aes128Gcm::new(GenericArray::from_slice(&key));
     let mut data = plaintext_resp_body.to_owned();
+    let mut aad = response_nonce.to_bytes()?;
```

Comment:
~~Does this include the length prefix? (I assume so since the test vectors pass.)~~

Yeah, looks that way, since ResponseNonce implements the custom serializer.